<launch>
  <!-- size of map, change the size inflate x, y, z according to your application -->
  <arg name="map_size_x" value="40.0"/>
  <arg name="map_size_y" value="40.0"/>
  <arg name="map_size_z" value=" 3.0"/>

  <!-- topic of your odometry such as VIO or LIO -->
  <arg name="odom_topic" default="/visual_slam/odom" />
  <arg name="depth_topic" default="/pcl_render_node/depth" />
  <arg name="max_yaw_offset_tolerance" default="30.0" />

  <!-- main algorithm params -->
  <include file="$(find ego_planner)/launch/advanced_param.xml">

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>
    <arg name="odometry_topic" value="$(arg odom_topic)"/>

    <!-- camera pose: transform of camera frame in the world frame -->
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- don't set cloud_topic if you already set these ones! -->
    <arg name="camera_pose_topic" value="/mavros/vision_pose/pose"/>
    <arg name="depth_topic" value="$(arg depth_topic)"/>

    <!-- topic of point cloud measurement, such as from LIDAR  -->
    <!-- don't set camera pose and depth, if you already set this one! -->
    <!--<arg name="cloud_topic" value="/pcl_render_node/cloud"/>-->

    <!-- intrinsic params of the depth camera -->
    <arg name="cx" value="64.5"/>
    <arg name="cy" value="48.5"/>
    <arg name="fx" value="64.0509852180153"/>
    <arg name="fy" value="64.0509852180153"/>
    <arg name="width" value="128"/>
    <arg name="height" value="96"/>

    <!-- maximum velocity and acceleration the drone will reach -->
    <arg name="max_vel" value="2.0" />
    <arg name="max_acc" value="3.0" />

    <!--always set to 1.5 times grater than sensing horizen-->
    <arg name="planning_horizon" value="7.5" /> 

    <!-- 1: use 2D Nav Goal to select goal  -->
    <!-- 2: use global waypoints below  -->
    <arg name="flight_type" value="1" />
    
    <!-- global waypoints -->
    <!-- It generates a piecewise min-snap traj passing all waypoints -->
    <arg name="point_num" value="5" />

    <arg name="point0_x" value="-15.0" />
    <arg name="point0_y" value="0.0" />
    <arg name="point0_z" value="1.0" />

    <arg name="point1_x" value="0.0" />
    <arg name="point1_y" value="15.0" />
    <arg name="point1_z" value="1.0" />

    <arg name="point2_x" value="15.0" />
    <arg name="point2_y" value="0.0" />
    <arg name="point2_z" value="1.0" />

    <arg name="point3_x" value="0.0" />
    <arg name="point3_y" value="-15.0" />
    <arg name="point3_z" value="1.0" />

    <arg name="point4_x" value="-15.0" />
    <arg name="point4_y" value="0.0" />
    <arg name="point4_z" value="1.0" />
    
  </include>

  <!-- trajectory server -->
  <node pkg="ego_planner" name="traj_server" type="traj_server" output="screen">
    <remap from="/position_cmd" to="planning/pos_cmd"/>

    <remap from="/odom_world" to="$(arg odom_topic)"/>
    <param name="traj_server/time_forward" value="1.0" type="double"/>
    <param name="traj_server/max_yaw_rate" value="30.0" type="double"/>
    <param name="traj_server/max_yaw_offset_tolerance" value="$(arg max_yaw_offset_tolerance)" />
    <param name="traj_server/tfTimeOut" value="0.01"/>
    <param name="traj_server/planning_frame" value="world"/>
    <param name="traj_server/planning_child_frame" value="fcu"/>

    <param name="traj_server/blind_goal_slow_down_dis" value="10.0"/>
    <!-- determins min blind speed -->
    <param name="traj_server/blind_goal_plan_ahead_dis_min" value="0.5"/>
    <!-- determins max blind speed -->
    <param name="traj_server/blind_goal_plan_ahead_dis_max" value="20.0"/>
    <!-- determins acc of blind flight -->
    <param name="traj_server/blind_goal_plan_ahead_dis_step_change_rate_max" value="1"/>
  </node>

  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
    <remap from="~odom" to="$(arg odom_topic)"/>        
    <remap from="~goal" to="/move_base_simple/goal"/>
    <remap from="~traj_start_trigger" to="/traj_start_trigger" />
    <param name="waypoint_type" value="manual-lonely-waypoint"/>    
  </node>

    <node pkg="odom_visualization" name="odom_visualization" type="odom_visualization" output="screen">
        <remap from="~odom" to="$(arg odom_topic)"/>
        <param name="color/a" value="1.0"/>
        <param name="color/r" value="0.0"/>
        <param name="color/g" value="0.0"/>
        <param name="color/b" value="1.0"/>
        <param name="covariance_scale" value="100.0"/>
        <param name="robot_scale" value="1.0"/>
  </node>
</launch>
